### PR TITLE
fix: replace '+' to '%2B' to handle page strs encode

### PR DESCRIPTION
**Pull request by imbajin**
When we get a `page` value from vertex/edge API like this
```javascript
GET xxx/graphs/hugegraph/graph/vertices?page&limit=3
```

We will get `page = xxxxx` to scan next page's values. however, when we get page value containis `+`, like this:
`ab4c+def123`

The url decoder(`@QueryParam`) will auto encode it to a space(refer to [W3C-HTML4.01](https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4.1)'s request), which leads to some error. So we condiser to 3 ways to solve it:
1. replace all `+` to `%2B` **before return** page to users (refer to[ W3C hex code](https://www.w3school.com.cn/tags/html_ref_urlencode.asp))
2. when param contains `+`, server will get a space, **then** we replace all space to `+`
3. use `getQueryString()` from HTTPRequest's URI, then split it and handle params.

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java [hunk 1]
```diff
@@ -118,6 +118,7 @@ private String writeIterator(String label, Iterator<?> iter,
                                             iter.getClass());
                 }
                 if (page != null) {
+                    page.replaceAll("[+]", "%2B");
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java:121]: prefer not transfer just when toJson

Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java:121]: are there any other illegal symbols?

Comment (imbajin) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java]: shall we keep encode `+` to `%2B`?  (just remove it now)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageState.java [hunk 2]
```diff
@@ -73,7 +73,8 @@ public String toString() {
 
     public static PageState fromString(String page) {
         E.checkNotNull(page, "page");
-        return fromBytes(toBytes(page));
+        String originalPage = page.replace(' ', '+');
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageState.java:76]: add some comment for why

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 3]
```diff
@@ -6667,6 +6668,37 @@ public void testQueryByPageWithInvalidPage() {
         });
     }
 
+
+    @Test
+    public void testQueryByPageWithSpecialBase64CharacterPage() {
+        // TODO: Not ready for commit, lack valid page contains '+' or '/'
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6674]: fiexed？

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 4]
```diff
@@ -6667,6 +6668,37 @@ public void testQueryByPageWithInvalidPage() {
...
+        // Assert decode space succees first
+        final String pageWithPlus = "5p2+"; // 松
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6676]: remove "// 松"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageState.java [hunk 5]
```diff
@@ -73,8 +75,13 @@ public String toString() {
 
     public static PageState fromString(String page) {
         E.checkNotNull(page, "page");
-        String originalPage = page.replace(' ', '+');
-        return fromBytes(toBytes(originalPage));
+        /*
+         * URLDecoder will auto decode '+' to space in url due to the request
+         * of HTML4, so we choose to replace the space to '+' after get it
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageState.java:80]: "after getting it."

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java (hunk 3 above)
Comment (imbajin) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java]: fixed now, it's hard to get a **valid page** contains '+' '/' '=' ....

**Review by Linary — APPROVED** [review]

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 6]
```diff
@@ -6667,6 +6669,37 @@ public void testQueryByPageWithInvalidPage() {
...
+                                     "AAAAAAAAAQ==";
+        Assert.assertNotNull(PageState.fromString(pageWith3Base64Chars));
+
+        byte[] decodePlus = PageState.fromString(pageWith3Base64Chars)
+                                     .position();
+        byte[] decodeSpace = PageState.fromString(pageWithSpace).position();
+
+        Assert.assertTrue(Arrays.equals(decodePlus, decodeSpace));
+        Assume.assumeTrue("Not support paging",
+                          storeFeatures().supportsQueryByPage());
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6690]: move to unittest

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 7]
```diff
@@ -6667,6 +6669,37 @@ public void testQueryByPageWithInvalidPage() {
...
+
+        HugeGraph graph = graph();
+        init100Books();
+
+        // Assert not throw exception when contains '+' or '/' or '=' or space
+        // Contains valid character '+' and '/' and '='
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6696]: use "/*" for multi-line comment or delete the first line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 8]
```diff
@@ -6667,6 +6669,37 @@ public void testQueryByPageWithInvalidPage() {
         });
     }
 
+    @Test
+    public void testQueryByPageWithSpecialBase64CharacterPage() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6673]: testQueryByPageWithSpecialBase64Chars

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 9]
```diff
@@ -6667,6 +6669,37 @@ public void testQueryByPageWithInvalidPage() {
...
+
+        HugeGraph graph = graph();
+        init100Books();
+
+        // Assert not throw exception when contains '+' or '/' or '=' or space
+        // Contains valid character '+' and '/' and '='
+        graph.traversal().V().has("~page", pageWith3Base64Chars)
+             .limit(10);
+        // Contains invalid base64 character ' ', will be replaced to '+'
+        graph.traversal().V().has("~page", pageWithSpace).limit(10);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6700]: add assert

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 10]
```diff
@@ -6667,6 +6669,37 @@ public void testQueryByPageWithInvalidPage() {
...
+        Assume.assumeTrue("Not support paging",
+                          storeFeatures().supportsQueryByPage());
+
+        HugeGraph graph = graph();
+        init100Books();
+
+        // Assert not throw exception when contains '+' or '/' or '=' or space
+        // Contains valid character '+' and '/' and '='
+        graph.traversal().V().has("~page", pageWith3Base64Chars)
+             .limit(10);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6698]: assert empty?

Comment (imbajin) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6698]: traversal is not empty, assert not null now

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 11]
```diff
@@ -6667,6 +6669,36 @@ public void testQueryByPageWithInvalidPage() {
         });
     }
 
+    @Test
+    public void testQueryByPageWithSpecialBase64Chars() {
+        final String pageWith3Base64Chars = "AAAAADsyABwAEAqI546LS6WW57unBgA" +
```
Comment (zhoney) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6674]: delete these lines, as already moved to unit test

**Review by zhoney — CHANGES_REQUESTED** [review]

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/PageStateTest.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                                     "B//// 8H////4alhxAZS8va6opcAKpklipAAQA" +
+                                     "AAAAAAAAAQ==";
+        Assert.assertNotNull(PageState.fromString(pageWith3Base64Chars));
+
+        byte[] decodePlus = PageState.fromString(pageWith3Base64Chars)
+                                     .position();
+        byte[] decodeSpace = PageState.fromString(pageWithSpace).position();
+
+        Assert.assertTrue(Arrays.equals(decodePlus, decodeSpace));
+    }
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/PageStateTest.java:48]: also add some test cases for other methods

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 13]
```diff
@@ -6667,6 +6669,36 @@ public void testQueryByPageWithInvalidPage() {
...
+                                            "EAAAAAPB////+8H////4alhxAZS8va6" +
+                                            "opcAKpklipAAQAAAAAAAAAAQ==";
+
+        final String pageWithSpace = "AAAAADsyABwAEAqI546LS6WW57unBgAEAAAAAP" +
+                                     "B//// 8H////4alhxAZS8va6opcAKpklipAAQA" +
+                                     "AAAAAAAAAQ==";
+
+        Assume.assumeTrue("Not support paging",
```
Comment (zhoney) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6682]: move to line 6674

**Review by zhoney — CHANGES_REQUESTED** [review]

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java (hunk 11 above)
Comment (imbajin) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java:6674]: keep them for tests below

**Review by zhoney — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Pull request merged**